packages/tokamak/contracts-bedrock/scripts/verify-predeploy-contracts.sh

```diff
@@ -0,0 +1,336 @@
+             ["4200000000000000000000000000000000000018"]=""  # ProxyAdmin
+             ["4200000000000000000000000000000000000019"]=""  # BaseFeeVault
+             ["420000000000000000000000000000000000001a"]=""  # L1FeeVault
+             ["4200000000000000000000000000000000000020"]=""  # SchemaRegistry
+             ["4200000000000000000000000000000000000021"]=""  # EAS
+             ["4200000000000000000000000000000000000486"]=""  # ETH (TOKAMAK)
+)
+
+IMPLEMENTATION_SLOT="0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc"
+
+echo "Getting contracts addresses..."
+for address in ${!contracts[@]}; do
+  QUERY=.alloc.\"$address\".storage.\"$IMPLEMENTATION_SLOT\"
+  contracts[${address}]=$(echo $(curl -fsSL $GENESIS_URL | jq $QUERY) | cut -c 28-67)
+
+  if [ -z ${contracts[${address}]} ]; then
+    contracts[${address}]=${address}
+  fi
+done
+echo "Successfully getting contracts addresses!"
+
+# Path of contracts
+BASE_PATH=$(cd $(dirname $0)/.. && pwd -P)
+PROXY_PATH=${BASE_PATH}/src/universal/Proxy.sol
+L2_TO_L1_MESSAGE_PASSER_PATH=${BASE_PATH}/src/L2/L2ToL1MessagePasser.sol
+DEPLOYER_WHITE_LIST_PATH=${BASE_PATH}/src/legacy/DeployerWhitelist.sol
+WTON_PATH=${BASE_PATH}/vendor/WTON.sol
+L2_CROSS_DOMAIN_MESSENGER_PATH=${BASE_PATH}/src/L2/L2CrossDomainMessenger.sol
+L2_STANDARD_BRIDGE_PATH=${BASE_PATH}/src/L2/L2StandardBridge.sol
+SEQUENCER_FEE_VAULT_PATH=${BASE_PATH}/src/L2/SequencerFeeVault.sol
+OPTIMISM_MINTABLE_ERC20_FACTORY_PATH=${BASE_PATH}/src/universal/OptimismMintableERC20Factory.sol
+L1_BLOCK_NUMBER_PATH=${BASE_PATH}/src/legacy/L1BlockNumber.sol
+GAS_PRICE_ORACLE_PATH=${BASE_PATH}/src/L2/GasPriceOracle.sol
+L1_BLOCK_PATH=${BASE_PATH}/src/L2/L1Block.sol
+LEGACY_MESSAGE_PASSER_PATH=${BASE_PATH}/src/legacy/LegacyMessagePasser.sol
+L2_ERC721_BRIDGE_PATH=${BASE_PATH}/src/L2/L2ERC721Bridge.sol
+OPTIMISM_MINTABLE_ERC721_FACTORY_PATH=${BASE_PATH}/src/universal/OptimismMintableERC721Factory.sol
+PROXY_ADMIN_PATH=${BASE_PATH}/src/universal/ProxyAdmin.sol
+BASE_FEE_VAULT_PATH=${BASE_PATH}/src/L2/BaseFeeVault.sol
+L1_FEE_VAULT_PATH=${BASE_PATH}/src/L2/L1FeeVault.sol
+SCHEMA_REGISTRY_PATH=${BASE_PATH}/src/EAS/SchemaRegistry.sol
+EAS_PATH=${BASE_PATH}/src/EAS/EAS.sol
+ETH_PATH=
+
+function run() {
+  verify_proxy
+  verify_L2ToL1MessagePasser
+  verify_DeployerWhitelist
+  #verify_WTON
+  verify_L2CrossDomainMessenger
+  verify_L2StandardBridge
+  verify_SequencerFeeVault
+  verify_OptimismMintableERC20Factory
+  verify_L1BlockNumber
+  verify_GasPriceOracle
+  verify_L1Block
+  verify_LegacyMessagePasser
+  verify_L2ERC721Bridge
+  verify_OptimismMintableERC721Factory
+  verify_ProxyAdmin
+  verify_BaseFeeVault
+  verify_L1FeeVault
+  verify_SchemaRegistry
+  verify_EAS
+  #verify_ETH
+}
+
+# Verify contract
+# $1 : compiler version
+# $2 : constructor args
+# $3 : contract address
+# $4 : contract path
+function verify_contract() {
+  forge verify-contract \
+    --chain-id $CHAIN_ID \
```

Comment:
Is there any verification logic for chain id? I was trying to test invalid chain id. but it works well.

Reply:
test result is here:

```bash
./verify-predeploy-contracts.sh   --chain-id 2312312323   --genesis-url https://tokamak-titan-canyon.s3.ap-northeast-2.amazonaws.com/titan-sepolia-test/genesis.json   --explorer-url https://explorer.titan-sepolia-test.tokamak.network
Getting contracts addresses...
Successfully getting contracts addresses!
Start verifying contract `0x4200000000000000000000000000000000000001` deployed on 2312312323
...
```

Reply:
thank you for your review!

i saw this review and In sepolia(etherscan), op-sepolia(etherscan), op-sepolia(blockscout), i created a test contract and tested it.

As a result, optimism blockscout has the same results as this review.

i think `forge verify-contract` command is not seem to require chainid in blockscout. it's only checks `--verifier-url` and verify contract.

in etherscan, It seems to require an api-key and then a chainid to check if that api-key is valid for that chain.

Since we're only using blockscout, should i remove the chain-id?

Reply:
Thanks for checking it out. The below is a my suggestion. 

I think we should not ask the user for the chain id as input, but use it as a constant in your code. (But this method is that we have to fix your code if we have change network.)  I've noticed that if we don't put the chain flag (`--chain-id or --chain`) in the verify-contract command, it shows up as mainnet, as shown below. 

```
Start verifying contract `0x4200000000000000000000000000000000000042` deployed on mainnet

Contract [src/governance/GovernanceToken.sol:GovernanceToken] "0x4200000000000000000000000000000000000042" is already verified. Skipping verification.
Start verifying contract `0xc0D3C0d3C0d3C0D3c0d3C0d3c0D3C0d3c0d30000` deployed on mainnet

Contract [src/legacy/LegacyMessagePasser.sol:LegacyMessagePasser] "0xc0D3C0d3C0d3C0D3c0d3C0d3c0D3C0d3c0d30000" is already verified. Skipping verification.
...
```

Reply:
thank you for checking again! i missed it that point.

my suggestion is when the user uses the script, they are explicitly asked for the network name (the script has a list of valid networks), and  hardcode the chainid in script.

This way, if a network is added in the future, the only thing to change is add the new chainid.

pseudocode example:
```bash
### run script
./verify_script.sh --network sepolia-test ...

### in script
NETWORK_NAME=$1

# we only add new network id in future
NETWORK_LIST = (
  "sepolia-test" : 111551115050,
  "mainnet" : 1010
)

for network in $NETWORK_LIST
  if network == NETWORK_NAME
    CHAIN_ID=NETWORK_LIST[${network}]
  fi
done

if [ -z CHAIN_ID ]
  echo "not vaild chainid"
  exit 1
fi
```

how about this way?

Reply:
Thanks for your agile feedback! It would be nice to show a list of supported networks in the main so that the user can select them.

Reply:
i fixed it!

```bash
./packages/tokamak/contracts-bedrock/scripts/verify-predeploy-contracts.sh --help                         
Verify L2 Pre-deploy contracts

Usage: /path/to/script/verify-predeploy-contracts.sh <ARGS>

Arguments:

  -n,    --network      (required)               Set target network
  -gURL, --genesis-url  (required)               Set genesis file for get contracts addresses
  -eURL, --explorer-url (required)               Block explorer URL to verify contracts(only blockscout)

Network list:

  titan-sepolia-test

Example:

  ./verify-predeploy-contracts.sh \
  --network titan-sepolia-test \
  --genesis-url https://tokamak-titan-canyon.s3.ap-northeast-2.amazonaws.com/titan-sepolia-test/genesis.json \
  --explorer-url https://explorer.titan-sepolia-test.tokamak.network
```